File: app/pricing/design_cp.py
"""Design CP calculation (spec §2–§3)."""

from app.pricing.models import CostAssumptions, DesignCPBreakdown, SlicerMetrics


def compute_effective_machine_time(batch_print_time_hr: float, units_per_bed: int) -> float:
    """Per-unit machine time once a plate is batched (spec §2).

    Example: a 4-unit plate printing in 5.5h → 1.375h per unit.
    """
    if units_per_bed < 1:
        raise ValueError("units_per_bed must be >= 1")
    if batch_print_time_hr <= 0:
        raise ValueError("batch_print_time_hr must be > 0")
    return round(batch_print_time_hr / units_per_bed, 4)


def compute_design_cp(
    metrics: SlicerMetrics, assumptions: CostAssumptions
) -> DesignCPBreakdown:
    """Total Design CP for a single unit (spec §3)."""
    filament = metrics.filament_g / 1000 * assumptions.filament_cost_per_kg
    purge = metrics.purge_g / 1000 * assumptions.filament_cost_per_kg
    electricity = metrics.electricity_kwh * assumptions.electricity_cost_per_unit
    machine = metrics.effective_machine_time_hr * assumptions.machine_hour_cost
    finishing = assumptions.finishing_labour
    consumables = assumptions.consumables

    subtotal = filament + purge + electricity + machine + finishing + consumables
    failure_provision = subtotal * assumptions.failure_pct
    design_cp = subtotal + failure_provision

    return DesignCPBreakdown(
        filament=round(filament, 2),
        purge=round(purge, 2),
        electricity=round(electricity, 2),
        machine=round(machine, 2),
        finishing=round(finishing, 2),
        consumables=round(consumables, 2),
        failure_provision=round(failure_provision, 2),
        subtotal=round(subtotal, 2),
        design_cp=round(design_cp, 2),
    )

File: internal/pricing/design_cp.go
package pricing

import "fmt"

// ComputeEffectiveMachineTime returns the per-unit print time after batching
// units_per_bed units on one bed, rounded to 4 dp. It mirrors
// compute_effective_machine_time in app/pricing/design_cp.py.
func ComputeEffectiveMachineTime(batchPrintTimeHr float64, unitsPerBed int) (float64, error) {
	if unitsPerBed < 1 {
		return 0, fmt.Errorf("units_per_bed must be at least 1")
	}
	if batchPrintTimeHr <= 0 {
		return 0, fmt.Errorf("batch_print_time_hr must be positive")
	}
	return roundHalfEven(batchPrintTimeHr/float64(unitsPerBed), 4), nil
}

// ComputeDesignCP builds the true per-unit manufacturing cost from slicer
// metrics and cost assumptions, plus a failure provision.
//
// The subtotal, failure provision and design CP are derived from the UNROUNDED
// components; only the emitted breakdown fields are rounded to 2 dp. This
// ordering is deliberate and matches the Python engine -- rounding the
// components first would let paisa-level error accumulate into the total.
func ComputeDesignCP(m SlicerMetrics, a CostAssumptions) DesignCPBreakdown {
	filament := m.FilamentG / 1000 * a.FilamentCostPerKg
	purge := m.PurgeG / 1000 * a.FilamentCostPerKg
	electricity := m.ElectricityKwh * a.ElectricityCostPerUnit
	machine := m.EffectiveMachineTimeHr * a.MachineHourCost
	finishing := a.FinishingLabour
	consumables := a.Consumables

	subtotal := filament + purge + electricity + machine + finishing + consumables
	failureProvision := subtotal * a.FailurePct
	designCP := subtotal + failureProvision

	return DesignCPBreakdown{
		Filament:         roundHalfEven(filament, 2),
		Purge:            roundHalfEven(purge, 2),
		Electricity:      roundHalfEven(electricity, 2),
		Machine:          roundHalfEven(machine, 2),
		Finishing:        roundHalfEven(finishing, 2),
		Consumables:      roundHalfEven(consumables, 2),
		FailureProvision: roundHalfEven(failureProvision, 2),
		Subtotal:         roundHalfEven(subtotal, 2),
		DesignCP:         roundHalfEven(designCP, 2),
	}
}
